language/stdlib/modules/Event.move

```diff
@@ -69,6 +75,189 @@ module Event {
+        define serialized_ehg_addr(ehg: EventHandleGenerator): vector<u8> {
+            LCS::serialize(ehg.addr)
+        }
+    }
+
+    /// ## Management of EventHandleGenerators
+
+    spec module {
+        /// Variable that is true whenever an `EventHandleGenerator` is destroyed.
+        /// This should never to true to preserve uniqueness of EventHandleGenerators.
+        global ehg_destroyed: bool;
+    }
+    spec struct EventHandleGenerator {
+        /// Updates the ehg_destroyed variable to true if an
+        /// `EventHandleGenerator` is ever unpacked.
+        invariant pack ehg_destroyed = ehg_destroyed;
+        invariant unpack ehg_destroyed = true;
+    }
+    spec schema EventHandleGeneratorNeverDestroyed {
+        /// **Informally:** No EventHandleGenerator should ever be destroyed.
+        invariant module !ehg_destroyed;
+    }
+    spec schema EventHandleGeneratorAtSameAddress {
+        /// **Informally:** An EventHandleGenerator should be located at `addr` and is never moved.
+        invariant module forall addr: address where initialized_ehg(addr): ehg(addr).addr == addr;
+    }
+    spec module {
+        /// Apply `EventHandleGeneratorNeverDestroyed` to all functions to ensure that the
+        /// `EventHandleGenerator`s are never destroyed.
+        /// Together with `CounterMonotonicallyIncreases`, this proves the uniqueness of the
+        /// `EventHandleGenerator` resource throughout transient states at each address.
+        /// Without this, the specification would allow an implementation to remove
+        /// an `EventHandleGenerator`, say `ehg`, and then have it generate a number
+        /// of events until the `ehg.counter` >= `old(ehg.counter)`. Violating the property
+        /// that an EventHandleGenerator should only be used to generate unique GUIDs.
+        ///
+        // BUG: Takes a long time (sometimes returns a precondition violation error)
+        // unless fresh_guid is excepted.
+        apply EventHandleGeneratorNeverDestroyed to * except fresh_guid;
+
+        /// Apply `EventHandleGeneratorAtSameAddress` to all functions to enforce
+        /// that all `EventHandleGenerator`s reside at the address they hold.
+        // BUG: Takes a long time unless fresh_guid is excepted.
+        apply EventHandleGeneratorAtSameAddress to * except fresh_guid;
+    }
+    spec fun publish_generator {
+        /// Creates a new `EventHandleGenerator` with an initial counter 0 and the
+        /// signer `account`'s address.
+        aborts_if exists<EventHandleGenerator>(Signer::get_address(account));
+        ensures global<EventHandleGenerator>(Signer::get_address(account))
+                    == EventHandleGenerator { counter: 0, addr: Signer::get_address(account) };
+    }
+
+    // Switch documentation context back to module level.
+    spec module {}
+
+    /// ## Uniqueness of EventHandleGenerator and Counter Monotonicity
+
+    spec schema CounterMonotonicallyIncreases {
```

Comment:
A more precise spec would be that the counter does not change unless new_event_handler is called (then it increases by 1).  However, I don't know if anything depends on this more precise property.